tokio-net/src/driver/reactor/sharded_slab/page/slot.rs

```diff
@@ -0,0 +1,105 @@
+use super::super::{Pack, Tid, RESERVED_BITS, WIDTH};
+use super::ScheduledIo;
+use crate::sync::{
+    atomic::{AtomicUsize, Ordering},
+    CausalCell,
+};
+
+#[derive(Debug)]
+pub(crate) struct Slot {
+    gen: AtomicUsize,
+    /// The offset of the next item on the free list.
+    next: CausalCell<usize>,
+    /// The data stored in the slot.
+    item: ScheduledIo,
+}
+
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub(crate) struct Generation {
+    value: usize,
+}
+
+impl Pack for Generation {
+    /// Use all the remaining bits in the word for the generation counter, minus
+    /// any bits reserved by the user.
+    const LEN: usize = (WIDTH - RESERVED_BITS) - Self::SHIFT;
+
+    type Prev = Tid;
+
+    #[inline(always)]
+    fn from_usize(u: usize) -> Self {
+        debug_assert!(u <= Self::BITS);
+        Self::new(u)
+    }
+
+    #[inline(always)]
+    fn as_usize(&self) -> usize {
+        self.value
+    }
+}
+
+impl Generation {
+    fn new(value: usize) -> Self {
+        Self { value }
+    }
+}
+
+impl Slot {
+    pub(super) fn new(next: usize) -> Self {
+        Self {
+            gen: AtomicUsize::new(0),
+            item: ScheduledIo::default(),
+            next: CausalCell::new(next),
+        }
+    }
+
+    #[inline(always)]
+    pub(super) fn get(&self, gen: Generation) -> Option<&ScheduledIo> {
+        let current = self.gen.load(Ordering::Acquire);
```

Comment:
Can you explain this? A guard w/ `load(Acquire)` raises flags in my mind.

Reply:
@carllerche do you mean that this needs to be an RMW in order to participate in a release sequence with `reset`, and would using a `fetch_add(0, Acquire)` work? Or does the ABA guard need to be sequentially consistent?

Reply:
As per our conversation offline, I agree that this is probably not correct. I wasn't able to write a test that fails due to this, but I've gone ahead and changed the generation counter to use `SeqCst` in 4017560